### ISSUE TITLE
mongoose-schema-extend

### ISSUE THREAD

**Issue by TomMettam** [szokodiakos/typegoose#117]
Hi there. Is there any way to use Typegoose with mongoose-schema-extend?

I can't use the mongoose discriminator because this only allows one level of inheritance.

**Comment by kaihaase-wd** [comment]
@TomMettam We extend / replace our models as follows:

**Extend**
```
import (User) from '...';
import * as mongoose from 'mongoose';
import { ModelType, prop } from 'typegoose';

// Extended user
export class ExtendedUser extends User {
	@prop()
	public hasAgreedToNewsletter?: boolean;
}

// Extend mongoose schema for user
export const UserModel: ModelType<ExtendedUser> = new ExtendedUser().getModelForClass(ExtendedUser, {
	existingMongoose: mongoose,
	schemaOptions: { collection: 'users' }
});
```

**Replace**
```
import (User as BasicUser) from '...'
import * as mongoose from 'mongoose';
import { ModelType, prop } from 'typegoose';

// Extended user
export class User extends BasicUser {
	@prop()
	public hasAgreedToNewsletter?: boolean;
}

// Extend mongoose schema for user
delete (<any>mongoose.connection).models['User'];
export const UserModel: ModelType<User> = new User().setModelForClass(User, {
	existingMongoose: mongoose,
	schemaOptions: { collection: 'users' }
});
```
Important for the replacement is to delete the old shema and to use the setModelForClass method (and not the getModelForClass method)

**Comment by TomMettam** [comment]
How does it know which model to load, in that case? There doesn't seem to be any discriminator key?

**Comment by kaihaase-wd** [comment]
@TomMettam That's a legitimate question! I optimized my comment and renamed UserModel to ExtendedUserModel. 

But this is basically a beauty correction, because it depends on which source is selected for the import:

import { UserModel } form 'user.model.ts' vs import { UserModel } from 'extended-user.model.ts'

resp. 

import { UserModel } form 'user.model.ts' vs import { ExtendedUserModel } from 'extended-user.model.ts'

**Comment by TomMettam** [comment]
Sorry, I think I should explain better..

With mongoose-schema-extend, if I search for a document by ID, it will automatically load the correct model based on the discriminator key (i.e if _type is ExtendedUser it will load the ExtendedUser model with all associated methods, if it's a regular user it'll just load the base model methods).

In simple terms: I can do a find on "UserModel" but it might return an ExtendedUserModel instead.

This is very very useful in my application, because I don't necessarily know what "subclass" a document is supposed to be before I load it. 

Is there any way to do this, or will I have to first load the model, then establish the type, then re-construct the model with the correct class?

**Comment by kaihaase-wd** [comment]
Unfortunately, there is no such automatism at the moment. I'd find that quite useful, though. Do you have a suggestion for the integration of mongoose-schema-extend?

**Comment by hasezoey** [comment]
this can be closed in favor of #332 

@Ben305

**Comment by hasezoey** [comment]
@Ben305 Bump to close
